app/__init__.py

```diff
@@ -300,3 +317,71 @@ def process_user_agent(user_agent_string):
         return "non-notify-user-agent"
     else:
         return "unknown"
+
+
+def setup_sqlalchemy_events(app):
+
+    TOTAL_DB_CONNECTIONS = Gauge(
+        'db_connection_total_connected',
+        'How many db connections are currently held (potentially idle) by the server',
+    )
+
+    TOTAL_CHECKED_OUT_DB_CONNECTIONS = Gauge(
+        'db_connection_total_checked_out',
+        'How many db connections are currently checked out by web requests',
+    )
+
+    DB_CONNECTION_OPEN_DURATION_SECONDS = Histogram(
+        'db_connection_open_duration_seconds',
+        'How long db connections are held open for in seconds',
+        ['method', 'host', 'path']
+    )
+
+    # need this or db.engine isn't accessible
+    with app.app_context():
+        @event.listens_for(db.engine, 'connect')
+        def connect(dbapi_connection, connection_record):
+            # connection first opened with db
+            TOTAL_DB_CONNECTIONS.inc()
```

Comment:
Optional: Add an if statement to only increment this and decrease on close if `has_request_context` is true as the metrics themselves are not still not exposed for celery apps etc

Reply:
hmmm, i don't think i'll do this right now as im not concerned about the actual act of writing metrics on boxes that don't expose things, i'm only concerned about errors introduced by metrics requiring access to variables they don't have (eg the `request.host` flask variable and stuff like that). i think the penalty of incrementing/decrementing a single gauge should be low enough that i'm not worried about it